### ISSUE TITLE
Add a disabled-by-default diagnostic for implicitly `Unknown`-specialized types used in annotations

### ISSUE THREAD

**Issue by ekohilas** [astral-sh/ty#2442]
### Question

Is there a way to make "ty" not pass for the following? (such as toggling a "strict" mode like "mypy" and "pyright")

```py
import re

def handle_pattern(m: re.Match) -> str:
    return m.string + 1
```

> re.Match is a generic class: it accepts a type parameter, in this case re.Match[str] or re.Match[bytes], to specify whether it represents a match over a unicode string or a byte string.

> If you don't specify a type parameter, it is assumed to be Any, so m: re.Match means m: re.Match[Any], and therefore m.string also has type Any. Just like any in TypeScript, Any in Python is "viral": it allows all operations, and the result of every operation is also Any.

> Always specify the parameter to re.Match (in most cases it's re.Match[str]). Both mypy and pyright will require this in strict mode, and they have more granular settings as well.

Source: https://decorator-factory.github.io/python-re-quiz/?quiz-pos=16_typing-match-any

### Version

0.0.11

**Comment by AlexWaygood** [comment]
We don't offer this diagnostic right now, but I think it's definitely an opt-in diagnostic that we should add at some point! I'm not actually sure that we have an issue open for it right now, so I'll repurpose this as a tracking issue for that.

**Comment by dhruvmanila** [comment]
Should this be part of https://github.com/astral-sh/ty/issues/527 ?

**Comment by oliverlambson** [comment]
+1 would really like this

**Comment by carljm** [comment]
This is also closely related to #622.

**Comment by jelle-openai** [comment]
Adding some keywords that would have helped me find this: disallow-any-generics (corresponding mypy option), default specialization.

**Comment by JelleZijlstra** [comment]
This is done, thanks @reginaldl!

I feel this should be on by default (maybe in a strict mode?), but that can be changed later.
